rs/pocket_ic_server/src/pocket_ic.rs

```diff
@@ -1621,43 +1616,20 @@ impl Operation for CallRequest {
                     .body(self.bytes.clone().into())
                     .unwrap();
 
-                let cancellation_token = CancellationToken::new();
-                let cancellation_token_clone = cancellation_token.clone();
-
-                // TODO: Allow parallel execution of V3 ingress messages.
-                //
-                // We are blocking the pic here, when in the CallRequest operation.
-                // This won't let us execute V3 ingress messages concurrently.
-                let resp = std::thread::scope(|s| {
-                    // We have to execute rounds for V3 calls, since the endpoint
-                    // waits for message to be executed and certified.
-                    if let CallRequestVersion::V3 = self.version {
-                        s.spawn(|| {
-                            while !cancellation_token_clone.is_cancelled() {
-                                for subnet in pic.subnets.read().unwrap().values() {
-                                    subnet.execute_round();
-                                }
-                                std::thread::sleep(EXECUTE_ROUND_INTERVAL);
-                            }
-                        });
-                    }
-
-                    let response = pic.runtime.block_on(svc.oneshot(request)).unwrap();
-                    cancellation_token.cancel();
-                    response
+                let fut: ApiResponse = Box::pin(async {
+                    let resp = svc.oneshot(request).await.unwrap();
+                    into_api_response(resp).await
                 });
+                let api_resp = fut.shared();
+                let service_task = pic.runtime.spawn(api_resp.clone());
 
-                OpOut::RawResponse((
-                    resp.status().into(),
-                    resp.headers()
-                        .iter()
-                        .map(|(name, value)| (name.as_str().to_string(), value.as_bytes().to_vec()))
-                        .collect(),
-                    pic.runtime
-                        .block_on(axum::body::to_bytes(resp.into_body(), usize::MAX))
-                        .unwrap()
-                        .to_vec(),
-                ))
+                // For the sake of determinism, we need to wait until one of
+                // `service_task` or `ingress_proxy_task` terminates:
+                // then all the state modifications have been performed
+                // and we can return from the operation.
+                while !service_task.is_finished() && !ingress_proxy_task.is_finished() {}
```

Comment:
Are these checks cheap?

Reply:
I don't know, but I'd like to not introduce a sleep to keep the update call as fast as possible. In any case, the operation computation is blocking so we're gonna block a thread in any case, right?

Reply:
I see they are from the task API, so I am going to assume it's fine.